### PR TITLE
Add token logic to BE

### PR DESCRIPTION
**Pull request by sophiebpeneva**

### CODE REVIEW

File: backend/controllers/error_controller.py [hunk 1]
```diff
@@ -95,7 +96,8 @@ def get_first_same_date_indexes_in_sublists(dates1, dates2, index1, index2):
     return get_first_same_date_indexes_in_sublists(dates1, dates2, index1, index2 + 1)
 
 
-class Error(Resource):
+class Error(ProtectedResource):
+
     def get(self, filename1, filename2):
         data1, _, _ = get_data(os.path.join(config.UPLOAD_FOLDER, secure_filename(filename1)))
```
Comment (kugelfish42) [inline backend/controllers/error_controller.py:102]: Didn't you change the filename convention or does the client make the switch?

File: backend/authentication/auth.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    }
+    jwt_token = jwt.encode(payload, TokenControllerSingleton.get_secret_key())
+    return jwt_token
+
+
+class ProtectedResource(Resource):
+    method_decorators = [auth_decorator]
+
+
+class Register(Resource):
```
[the remark this answers is not in the document]
Comment (sophiebpeneva) [reply backend/authentication/auth.py:34]: On opening the web app, the client hits this endpoint with a random generated username

File: backend/controllers/error_controller.py (hunk 1 above)
Comment (sophiebpeneva) [reply backend/controllers/error_controller.py:102]: In the file_controller I include the new filename in the response and that's what the client sends in this request, so yes the client makes the switch

File: backend/controllers/file_controller.py [hunk 3]
```diff
@@ -1,26 +1,35 @@
...
         if file.filename == '':
             abort(400, 'Filename cannot be empty')
         if allowed_file(file.filename):
-            filename = secure_filename(file.filename)
+            filename = request.username + str(int(time.time()))
             file.save(os.path.join(config.UPLOAD_FOLDER, filename))
-            return {'response': 'File uploaded successfully'}
+            user_files = request.filenames
+            user_files.append(filename)
+            jwt_token = issue_token(request.username, user_files)
```
[the remark this answers is not in the document]
Comment (sophiebpeneva) [reply backend/controllers/file_controller.py:27]: user_files is a list. So if the BE gets a token with a paylaod {username:'sophie', filenames:['file']} , I save the ['file'] list in request.filenames and here I append the the filename on line 26 so the new token would be with a payload {username:'sophie', filenames:['file', 'new_file']}

File: backend/middlewares/access_validator.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+
+def validate_token(request, secret_key):
+    auth_header = request.headers.get('Authorization')
+    if not auth_header:
+        abort(403, 'Bearer token is necessary for authorization')
+    auth_token = auth_header.split(" ")[1]
+    try:
+        payload = jwt.decode(auth_token, secret_key)
+        request.filenames = payload['filenames']
```
[the remark this answers is not in the document]
Comment (sophiebpeneva) [reply backend/middlewares/access_validator.py:12]: renamed method

File: backend/controllers/file_controller.py [hunk 5]
```diff
@@ -1,26 +1,35 @@
...
+class File(ProtectedResource):
     def post(self):
         if 'uploaded_data' not in request.files:
             abort(400, 'Uploaded_data is required for the request')
         file = request.files['uploaded_data']
         if file.filename == '':
             abort(400, 'Filename cannot be empty')
         if allowed_file(file.filename):
-            filename = secure_filename(file.filename)
+            filename = request.username + str(int(time.time()))
```
[the remark this answers is not in the document]
Comment (sophiebpeneva) [reply backend/controllers/file_controller.py]: done

File: backend/authentication/token_controllerr.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import random
+import string
+import config
+
+
+class TokenControllerSingleton:
```
[the remark this answers is not in the document]
Comment (sophiebpeneva) [reply backend/authentication/token_controllerr.py]: done

File: backend/controllers/error_controller.py (hunk 1 above)
Comment (kugelfish42) [reply backend/controllers/error_controller.py:102]: Such kind of subtle and non-obvious conventions should probably be documented some in the code - ideally both in client and server code where this needs to happen.

**Review by kugelfish42 — APPROVED** [review]

**Pull request merged**